Marketplace/src/main/java/com/marketplace/controller/category/CategoryController.java

```diff
@@ -0,0 +1,82 @@
+package com.marketplace.controller.category;
+
+import com.marketplace.repository.category.ProductQuery;
+import com.marketplace.service.category.Category;
+import com.marketplace.service.category.CategoryService;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.servlet.mvc.support.RedirectAttributes;
+
+import java.security.Principal;
+
+@Controller
+@RequestMapping("categories")
+public class CategoryController {
+    private static final int NUMBER_OF_ENTRIES = 100;
+    private final CategoryService service;
+
+    public CategoryController(CategoryService service) {
+        this.service = service;
+    }
+
+    @GetMapping("{parentId}/new")
+    public String addCategory(@PathVariable long parentId, Model model) {
+        model.addAttribute("category", new Category(parentId));
+        return "new-category";
+    }
+
+    @PostMapping
+    public String addCategory(@ModelAttribute Category category) {
+        service.addCategory(category);
+        return "redirect:/categories" + category.getParentId();
+    }
+
+    @PutMapping("{categoryId}")
+    public String removeCategory(@PathVariable long categoryId) {
+        service.removeCategory(categoryId);
+        return "redirect:/categories";
+    }
+
+    @GetMapping(value = {"", "{categoryId}"})
+    public String getCategories(@PathVariable(required = false) Long categoryId,
```

Comment:
GET /categories and GET categories/{id} should be different endpoints

Reply:
This request returns subcategories of a category. When categoryId is not specified - it means that user looks for 'root' categories, which doesn't have parent categories